readthedocs/core/permissions.py

```diff
@@ -1,18 +1,86 @@
-
 """Objects for User permission checks."""
+from django.conf import settings
 from django.contrib.auth.models import User
 from django.db.models import Q
 
 from readthedocs.core.utils.extend import SettingsOverrideObject
+from readthedocs.organizations.constants import ADMIN_ACCESS, READ_ONLY_ACCESS
 
 
 class AdminPermissionBase:
 
     @classmethod
-    def admins(cls, obj):
+    def projects(cls, user, admin=False, member=False):
+        """
+        Return all the projects the user has access to as ``admin`` or ``member``.
+
+        If `RTD_ALLOW_ORGANIZATIONS` is enabled
+        this function takes SSO into consideration, it includes:
+
+        - Projects where the user has access to via SSO.
+        - Projects where the user has access via Team,
+          if SSO is not enabled for the organization of that team.
+
+        .. note::
+
+           SSO is taken into consideration,
+           but isn't implemented in .org yet.
+
+        :param user: user object to filter projects
+        :type user: django.contrib.admin.models.User
+        :param bool admin: include projects where the user has admin access to the project
+        :param bool member: include projects where the user has read access to the project
+        """
         from readthedocs.projects.models import Project
+        projects = Project.objects.none()
+        if not user or not user.is_authenticated:
+            return projects
+
+        if not settings.RTD_ALLOW_ORGANIZATIONS:
+            # All users are admin and member of a project
+            # when we aren't using organizations.
+            return user.projects.all()
+
+        if admin:
+            # Project Team Admin
+            admin_teams = user.teams.filter(access=ADMIN_ACCESS)
+            for team in admin_teams:
+                if not cls._has_sso_enabled(team.organization):
+                    projects |= team.projects.all()
+
+            # Org Admin
+            for org in user.owner_organizations.all():
+                if not cls._has_sso_enabled(org):
+                    # Do not grant admin access on projects for owners if the
+                    # organization has SSO enabled with Authorization on the provider.
+                    projects |= org.projects.all()
+
+            projects |= cls._get_projects_for_sso_user(user, admin=True)
+
+        if member:
+            # Project Team Member
+            member_teams = user.teams.filter(access=READ_ONLY_ACCESS)
+            for team in member_teams:
+                if not cls._has_sso_enabled(team.organization):
+                    projects |= team.projects.all()
+
+            projects |= cls._get_projects_for_sso_user(user, admin=False)
+
+        return projects
+
+    @classmethod
+    def _has_sso_enabled(cls, organization):
+        return False
+
+    @classmethod
+    def _get_projects_for_sso_user(cls, user, admin=False):
+        from readthedocs.projects.models import Project
+        return Project.objects.none()
```

Comment:
don't really like having these as placeholders, but was this or repeat the whole thing in .com or block this till we move the vcs integration.

Reply:
Why don't bring `CorporateAdminPermission._get_projects_for_sso_user` here and use the setting that you just added `RTD_ALLOW_ORGANIZATIONS` to decide what to do?

Reply:
In any case, if we are already planning to change the content of this method, I'd like to see a `TODO:` list explaining the plan here and what's the code we need to bring here.

Reply:
>  Why don't bring CorporateAdminPermission._get_projects_for_sso_user here and use the setting that you just added RTD_ALLOW_ORGANIZATIONS to decide what to do?

It has references to models that we don't have here.

> I'd like to see a TODO: list explaining the plan here and what's the code we need to bring here.

This is already overriden in .com.

Reply:
Now that we have the models in community, we can move the code from corporate here and use the `RTD_ALLOW_ORGANIZATIONS` as suggested, right? That way we will have all the code in the same place and once we enable organization in community it will just work.